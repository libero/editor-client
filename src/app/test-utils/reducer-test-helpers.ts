import { EditorState } from 'prosemirror-state';
import { Schema } from 'prosemirror-model';

import { Manuscript } from 'app/models/manuscript';
import { getInitialHistory, getLoadableStateSuccess } from 'app/utils/state.utils';
import { ManuscriptHistoryState } from 'app/store';

export function givenState(overrides: Partial<Manuscript>): ManuscriptHistoryState {
  return getLoadableStateSuccess(
    getInitialHistory({
      title: createDummyEditorState(),
      abstract: createDummyEditorState(),
      authors: [],
      affiliations: [],
      keywordGroups: {},
      articleInfo: {
        dtd: '',
        articleType: '',
        articleDOI: ''
      },
      references: [],
      ...overrides
    })
  );
}

export function createDummyEditorState(): EditorState {
  return EditorState.create({
    schema: new Schema({
      nodes: {
        doc: { content: 'inline*' },
        text: { inline: true, group: 'inline' }
      }
    }),
    plugins: []
  });
}
