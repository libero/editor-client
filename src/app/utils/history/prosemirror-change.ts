import { Transaction } from 'prosemirror-state';
import { Manuscript } from 'app/types/manuscript';
import { get, set } from 'lodash';
import { Change } from 'app/utils/history/change';
import { cloneManuscript } from 'app/utils/state.utils';

export class ProsemirrorChange implements Change {
  constructor(private path: string, private transaction: Transaction) {}

  get isEmpty(): boolean {
    return !this.transaction.docChanged;
  }

  applyChange(manuscript: Manuscript): Manuscript {
    const editorState = get(manuscript, this.path);
    return set(cloneManuscript(manuscript), this.path, editorState.apply(this.transaction));
  }

  rollbackChange(manuscript: Manuscript): Manuscript {
    const invertedSteps = this.transaction.steps.map((step, index) => {
      return step.invert(this.transaction.docs[index]);
    });

    const editorState = get(manuscript, this.path);
    const rollbackTransaction = editorState.tr;
    invertedSteps.reverse().forEach((step) => rollbackTransaction.maybeStep(step));

    return set(cloneManuscript(manuscript), this.path, editorState.apply(rollbackTransaction));
  }
}
