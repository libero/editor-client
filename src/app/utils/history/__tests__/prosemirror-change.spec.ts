import { givenState } from '../../../test-utils/reducer-test-helpers';
import { ProsemirrorChange } from '../prosemirror-change';

describe('ProsemirrorChange', () => {
  const manuscript = givenState({}).data.present;

  it('should apply prosemirror change', () => {
    const prosemirrorChange = new ProsemirrorChange('body', manuscript.body.tr.insertText('test content'));

    const updatedManuscript = prosemirrorChange.applyChange(manuscript);
    expect(updatedManuscript.body.doc.textContent).toBe('test content');
  });

  it('should revert prosemirror change', () => {
    const prosemirrorChange = new ProsemirrorChange('body', manuscript.body.tr.insertText('test content'));
    const updatedManuscript = prosemirrorChange.applyChange(manuscript);

    expect(prosemirrorChange.rollbackChange(updatedManuscript)).toEqual(manuscript);
  });

  it('should indicate empty change', () => {
    const prosemirrorChange = new ProsemirrorChange('body', manuscript.body.tr);
    expect(prosemirrorChange.isEmpty).toBeTruthy();
  });

  it('should check if path is affected', () => {
    const change = new ProsemirrorChange('body', manuscript.body.tr);
    expect(change.isPathAffected(/^body$/)).toBeTruthy();
  });

  it('should check if path is affected but regex is invalid', () => {
    const change = new ProsemirrorChange('body', manuscript.body.tr);
    expect(change.isPathAffected(/^wibble$/)).toBeFalsy();
  });

  it('should serialize to JSON', () => {
    const change = new ProsemirrorChange('body', manuscript.body.tr.insertText('test content'));

    expect(change.toJSON()).toEqual({
      path: 'body',
      timestamp: expect.any(Number),
      type: 'prosemirror',
      transactionSteps: [
        {
          from: 0,
          slice: { content: [{ type: 'text', text: 'test content' }] },
          stepType: 'replace',
          to: 0
        }
      ]
    });
  });

  it('should deserialize from JSON', () => {
    const JSONObject = {
      path: 'body',
      timestamp: 1610979099826,
      transactionSteps: [
        {
          stepType: 'replace',
          from: 0,
          to: 0,
          slice: {
            content: [
              {
                type: 'text',
                text: 'This text is sent from the server'
              }
            ]
          }
        }
      ],
      type: 'prosemirror'
    };

    const change = ProsemirrorChange.fromJSON(JSONObject);

    expect(change).toMatchSnapshot();
    expect(change).toBeInstanceOf(ProsemirrorChange);
  });
});
