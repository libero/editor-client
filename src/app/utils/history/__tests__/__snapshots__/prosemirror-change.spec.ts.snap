// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProsemirrorChange should deserialize from JSON 1`] = `
Object {
  "path": "body",
  "timestamp": 1610979099826,
  "transactionSteps": Array [
    Object {
      "from": 0,
      "slice": Object {
        "content": Array [
          Object {
            "text": "This text is sent from the server",
            "type": "text",
          },
        ],
      },
      "stepType": "replace",
      "to": 0,
    },
  ],
  "type": "prosemirror",
}
`;
